c++/test/TestPredicateLeaf.cc

```diff
@@ -0,0 +1,827 @@
+    double min, double max, bool hasNull = false) {
+    proto::ColumnStatistics colStats;
+    colStats.set_hasnull(hasNull);
+    colStats.set_numberofvalues(10);
+
+    proto::DoubleStatistics * doubleStats = colStats.mutable_doublestatistics();
+    doubleStats->set_minimum(min);
+    doubleStats->set_maximum(max);
+    return colStats;
+  }
+
+  static proto::ColumnStatistics createDecimalStats(
+    Decimal min, Decimal max, bool hasNull = false) {
+    proto::ColumnStatistics colStats;
+    colStats.set_hasnull(hasNull);
+    colStats.set_numberofvalues(10);
+
+    proto::DecimalStatistics * decimalStats = colStats.mutable_decimalstatistics();
+    decimalStats->set_minimum(min.toString(true));
+    decimalStats->set_maximum(max.toString(true));
+    return colStats;
+  }
+
+  static proto::ColumnStatistics createDateStats(
+    int32_t min, int32_t max, bool hasNull = false) {
+    proto::ColumnStatistics colStats;
+    colStats.set_hasnull(hasNull);
+    colStats.set_numberofvalues(10);
+
+    proto::DateStatistics * dateStats = colStats.mutable_datestatistics();
+    dateStats->set_minimum(min);
+    dateStats->set_maximum(max);
+    return colStats;
+  }
+
+  static proto::ColumnStatistics createTimestampStats(
+    int64_t min, int64_t max, bool hasNull = false) {
+    proto::ColumnStatistics colStats;
+    colStats.set_hasnull(hasNull);
+    colStats.set_numberofvalues(10);
+
+    proto::TimestampStatistics * tsStats = colStats.mutable_timestampstatistics();
+    tsStats->set_minimumutc(min);
+    tsStats->set_maximumutc(max);
+    return colStats;
+  }
+
+  static proto::ColumnStatistics createStringStats(
+    std::string min, std::string max, bool hasNull = false) {
+    proto::ColumnStatistics colStats;
+    colStats.set_hasnull(hasNull);
+    colStats.set_numberofvalues(10);
+
+    proto::StringStatistics * strStats = colStats.mutable_stringstatistics();
+    strStats->set_minimum(min);
+    strStats->set_maximum(max);
+    return colStats;
+  }
+
+  static TruthValue evaluate(const PredicateLeaf& pred,
+                             const proto::ColumnStatistics& pbStats,
+                             const BloomFilter * bf = nullptr) {
+    return pred.evaluate(WriterVersion_ORC_135, pbStats, bf);
+  }
+
+  TEST(TestPredicateLeaf, testPredEvalWithColStats) {
+    PredicateLeaf pred0(
+      PredicateLeaf::Operator::NULL_SAFE_EQUALS,
+      PredicateType::BOOLEAN,
+      "x",
+      Literal(true));
+    EXPECT_EQ(TruthValue::YES_NO,
+              evaluate(pred0, createBooleanStats(10, 10)));
+    EXPECT_EQ(TruthValue::NO,
+              evaluate(pred0, createBooleanStats(10, 0)));
+
+    PredicateLeaf pred1(
+      PredicateLeaf::Operator::NULL_SAFE_EQUALS,
+      PredicateType::BOOLEAN,
+      "x",
+      Literal(false));
+    EXPECT_EQ(TruthValue::NO,
+              evaluate(pred1, createBooleanStats(10, 10)));
+    EXPECT_EQ(TruthValue::YES_NO,
+              evaluate(pred1, createBooleanStats(10, 0)));
+
```

Comment:
It'd be better to also test "x <=> NULL" (x NULL_SAFE_EQUALS to NULL).

Reply:
Done. I have resolved all issues. Can you take a look again? Thanks! @stiga-huang